### PR TITLE
DCP2-248 - bulid HTML/PDF artifacts

### PR DESCRIPTION
**Pull request by respinos**
Adds:

- `lib/um_arclight/package/generator.rb` to take index data and generate complete collection documents
- generates a usable HTML file served by `/html` (with current assets)
- generates a PDF served by `/pdf`
- some rake tasks!

```
# generate by EADID
bundle exec rake arclight:generate_pdf EADID=umich-bhl-2014150

# queue by REPOSITORY_ID, although we don't use this yet because of umich-bhl-2014150
bundle exec rake arclight:queue REPOSITORY_ID=bhl
```

### CODE REVIEW

**Comment by respinos** [comment]
Geez, and this was the cleaner branch :-/

File: lib/um_arclight/package/generator.rb [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+
+        File.open(output_filename, 'w') do |f|
+          f.puts doc.serialize
+        end
+      end
+
+      def build_pdf
+        # build the source in tmp
+        FileUtils.mkdir_p(working_path_name)
+        Dir.chdir(working_path_name)
```
Comment (botimer) [inline lib/um_arclight/package/generator.rb:80]: I guess the other pieces are dependent on cwd... We'll probably want to leak test this for threading issues, though it should always be running in a single-threaded context. There might be some potential for doing this in a block, like some of the tempfile and popen3 constructs, to make the "working dir-ness" of this whole operation more explicit.

File: lib/um_arclight/package/generator.rb [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+
+        puts "UM-Arclight generate package: #{collection.id} : puppeteer render (in #{elapsed_time.round(3)} secs)."
+      end
+
+      private
+
+      def generate_output_filename(ext)
+        filename = "#{DulArclight.finding_aid_data}/pdf/#{collection.repository_id}/#{collection.id}#{ext}"
+        filename = File.join(Rails.root, filename) if filename.start_with?('./')
+        filename
```
Comment (botimer) [inline lib/um_arclight/package/generator.rb:137]: This is another opportunity for using Pathname, which I've found both convenient and less prone to weird space and slash errors. Not needed now; an idea for later improvement.

**Review by botimer — APPROVED** [review]
There are a few areas I think we can improve over time, but no show-stoppers here. A lot of good stuff figured out! 🚢

File: app/jobs/package_finding_aid_job.rb [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+require 'um_arclight/errors'
+require 'um_arclight/package/generator'
+
+# Job to queue packaging
+class PackageFindingAidJob < ApplicationJob
+  queue_as :index
+
+  def perform(identifier)
+    artifact = UmArclight::Package::Generator.new identifier: identifier
```
[the remark this answers is not in the document]
Comment (respinos) [reply app/jobs/package_finding_aid_job.rb:11]: It was a struggle. I'm not super thrilled with landing on `Package` either.

File: lib/tasks/generate.rake [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+namespace :arclight do
+  artifact = nil
+
+  desc 'Generate packages for indexed finding aids via background jobs'
+  task generate_enqueue: :environment do
+    repository_ssm = nil
+    if ENV['REPOSITORY_ID']
+      repository_config = Arclight::Repository.find_by(slug: ENV['REPOSITORY_ID'])
+      repository_ssm = repository_config.name
+    end
```
[the remark this answers is not in the document]
Comment (respinos) [reply lib/tasks/generate.rake:15]: Probably!

**Pull request merged**